DPPP/BDABuffer.cc

```diff
@@ -0,0 +1,216 @@
+      }
+      if (fields.flags_) {
+        flags_.reserve(remaining_capacity_);
+      }
+      if (fields.weights_) {
+        weights_.reserve(remaining_capacity_);
+      }
+      if (fields.full_res_flags_) {
+        full_res_flags_.reserve(remaining_capacity_);
+      }
+    }
+
+    // When copying the memory pools in this copy-constructor, the capacity
+    // of the new memory pools will equal their size. There is therefore
+    // no remaining capacity in the new copy.
+    BDABuffer::BDABuffer(const BDABuffer& other)
+    : data_(other.data_)
+    , flags_(other.flags_)
+    , weights_(other.weights_)
+    , full_res_flags_(other.full_res_flags_)
+    , rows_()
+    , original_capacity_(0)
+    , remaining_capacity_(0)
+    {
+      // Copy rows but set their iterators to the new memory pools.
+      auto data_it = data_.begin();
+      auto flags_it = flags_.begin();
+      auto weights_it = weights_.begin();
+      auto full_res_flags_it = full_res_flags_.begin();
+
+      rows_.reserve(other.rows_.size());
+      for (const auto& row : other.rows_) {
+        rows_.emplace_back(row.time_,
+                           row.interval_,
+                           row.row_nr_,
+                           row.baseline_nr_,
+                           row.n_channels_,
+                           row.n_correlations_,
+                           data_it,
+                           flags_it,
+                           weights_it,
+                           full_res_flags_it,
+                           row.uvw_);
+
+        const std::size_t kDataSize = row.n_channels_ * row.n_correlations_;
+        data_it += kDataSize;
+        flags_it += kDataSize;
+        weights_it += kDataSize;
+        full_res_flags_it += kDataSize;
+        original_capacity_ += kDataSize;
+      }
+    }
+
+    void BDABuffer::Clear()
+    {
+      data_.clear();
+      flags_.clear();
+      weights_.clear();
+      full_res_flags_.clear();
+      rows_.clear();
+      remaining_capacity_ = original_capacity_;
+    }
+
+    std::size_t BDABuffer::GetNumberOfElements() const
+    {
+      return original_capacity_ - remaining_capacity_;
+    }
+
+    bool BDABuffer::AddRow(double time,
+                           double interval,
+                           rownr_t row_nr,
+                           std::size_t baseline_nr,
+                           std::size_t n_channels,
+                           std::size_t n_correlations,
+                           const std::complex<float>* const data,
+                           const bool* const flags,
+                           const float* const weights,
+                           const bool* const full_res_flags,
+                           const double* const uvw)
+    {
+      if (!rows_.empty() && TimeIsLessEqual(time + interval, rows_.back().time_)) {
+        throw std::invalid_argument("Rows are not properly ordered");
+      }
+
+      const std::size_t kDataSize = n_channels * n_correlations;
```

Comment:
To me it is a little bit confusing to name this like a formal constant. Google Style says "Variables declared constexpr or const, **and whose value is fixed for the duration of the program**, are named with a leading "k" followed by mixed case" -- since this is within a function, which is called multiple times maybe with different n_channels, I think this therefore does not satisfy that requirement, in line with how I would interpret a constant.

Reply:
When I read the style guide, I agree that kXxxx is only for fixed constants, not for 'variable' constants.